### PR TITLE
Add support for custom testnets

### PR DESCRIPTION
**Pull request by cbermudez97**
## Changes:
- Add support for custom testnets
## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests?**

- [ ] Yes
- [x] No

### CODE REVIEW

File: templates/services/docker-compose_base.tmpl [hunk 1]
```diff
@@ -19,13 +19,13 @@ services:
       - -relay-check
       - -relays
       - ${RELAY_URL}{{end}}
-  {{if or .CcCustomCfg .VlCustomCfg}}config_consensus:
+{{if or .CustomCfgDownload }}  config_consensus:
```
Comment (AntiD2ta) [inline templates/services/docker-compose_base.tmpl:22]:
```suggestion
{{if .CustomCfgDownload }}  config_consensus:
```

Comment (AntiD2ta) [reply templates/services/docker-compose_base.tmpl:22]: good one with removing the need for the or

File: templates/services/merge/execution/besu.tmpl [hunk 2]
```diff
@@ -24,7 +24,8 @@
     - --sync-mode=X_SNAP
     - --data-storage-format=BONSAI
     - --network={{if .SplittedNetwork}}${EL_NETWORK}{{else}}${NETWORK}{{end}}
```
Comment (AntiD2ta) [inline templates/services/merge/execution/besu.tmpl:26]: Add `--genesis-file` flag. This flag can't be used with `--network` at the same time

File: templates/services/merge/execution/erigon.tmpl [hunk 3]
```diff
@@ -27,7 +27,8 @@
     - --http.addr=0.0.0.0
     - --http.port={{.ElApiPort}}
     - --txpool.disable
-    - --chain={{if .SplittedNetwork}}${EL_NETWORK}{{else}}${NETWORK}{{end}}
+    - --chain={{if .SplittedNetwork}}${EL_NETWORK}{{else}}${NETWORK}{{end}}{{if .ECBootnodes}}
```
Comment (AntiD2ta) [inline templates/services/merge/execution/erigon.tmpl:30]: Research if erigon allow you to specify a custom genesis file

**Review by AntiD2ta — CHANGES_REQUESTED** [review]
LGTM, good refactor of the check variables as well. Minor changes requested

Also please update the Changelog

File: templates/services/merge/execution/erigon.tmpl (hunk 3 above)
Comment (cbermudez97) [reply templates/services/merge/execution/erigon.tmpl:30]: Similar to Geth in that aspect. The workflow to use a custom testnet doesn't match with the Sedge uses.

File: internal/pkg/generate/generate_scripts.go [hunk 4]
```diff
@@ -214,8 +264,14 @@ func generateDockerComposeScripts(gd GenerationData) (dockerComposePath string,
 		return "", err
 	}
 
+	// Check for CC Bootstrap nodes
```
[the remark this answers is not in the document]
Comment (cbermudez97) [reply internal/pkg/generate/generate_scripts.go:267]: Bootnode is the short version 😅

File: cli/cli.go [hunk 5]
```diff
@@ -136,6 +143,14 @@ func CliCmd(cmdRunner commands.CommandRunner, prompt prompts.Prompt, serviceMana
 	flags.clExtraFlags = cmd.Flags().StringArray("cl-extra-flag", []string{}, "Additional flag to configure the consensus client service in the generated docker-compose script. Example: 'sedge cli --cl-extra-flag \"<flag1>=value1\" --cl-extra-flag \"<flag2>=\\\"value2\\\"\"'")
 	flags.vlExtraFlags = cmd.Flags().StringArray("vl-extra-flag", []string{}, "Additional flag to configure the validator client service in the generated docker-compose script. Example: 'sedge cli --vl-extra-flag \"<flag1>=value1\" --vl-extra-flag \"<flag2>=\\\"value2\\\"\"'")
 	cmd.Flags().StringVar(&flags.logging, "logging", "json", fmt.Sprintf("Docker logging driver used by all the services. Set 'none' to use the default docker logging driver. Possible values: %v", configs.ValidLoggingFlags()))
+	cmd.Flags().StringVar(&flags.customTTD, "custom-ttd", "", "Custom Terminal Total Difficulty to use for execution and consensus clients")
```
[the remark this answers is not in the document]
Comment (cbermudez97) [reply cli/cli.go:146]: Both clients uses TTD to override the default ones. In both cases they change their behavior depending if the merge have occurred or not.

File: internal/pkg/env/check_variable_test.go [hunk 6]
```diff
@@ -124,52 +124,3 @@ func TestCheckVariable(t *testing.T) {
 		})
 	}
 }
-
-func TestCheckVariableBase(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (cbermudez97) [reply internal/pkg/env/check_variable_test.go:128]: TTD no longer uses that function. Yet it is true that the function is still being used, will add it back with different test cases.

File: internal/pkg/generate/generate_scripts.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/generate/generate_scripts.go:267]: Bootstrap can be misleading, please change it to Bootnode

File: cli/cli.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply cli/cli.go:146]: Done 68397a4f1c44fa1e8272b4358f6e732b7c2eba52

File: internal/utils/file_url_utils.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+func GetUrlOrPathContent(src string) (string, error) {
+	urlContent := ""
+	fileContent := ""
+
+	err := HandleUrlOrPath(
+		src,
+		func(url string) error {
+			resp, err := http.Get(url)
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/utils/file_url_utils.go]: Done [500e26a](https://github.com/NethermindEth/sedge/pull/196/commits/500e26ab59c50044d106bf257f814ad52b1680d3)

File: internal/utils/file_url_utils.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	destFile, err := os.Create(dest)
+	if err != nil {
+		return fmt.Errorf(configs.ErrorCreatingFile, dest, err)
+	}
+	defer destFile.Close()
+
+	return HandleUrlOrPath(
+		src,
+		func(url string) error {
+			resp, err := http.Get(url)
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/utils/file_url_utils.go]: Done [500e26a](https://github.com/NethermindEth/sedge/pull/196/commits/500e26ab59c50044d106bf257f814ad52b1680d3)

File: internal/pkg/env/check_variable_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/env/check_variable_test.go:128]: Done [d8388f4](https://github.com/NethermindEth/sedge/pull/196/commits/d8388f4f7e53892e861eebd604905d66a7108d54)

File: internal/pkg/generate/generate_scripts.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/generate/generate_scripts.go]: Done [8289795](https://github.com/NethermindEth/sedge/pull/196/commits/8289795e99bd5e630be4757ca2c99024d31442de)

File: internal/pkg/generate/generate_scripts.go [hunk 9]
```diff
@@ -180,27 +188,69 @@ func generateDockerComposeScripts(gd GenerationData) (dockerComposePath string,
...
 	if err != nil {
 		return "", err
 	}
+	if vlRemoteDpl { // Have to download custom configs
+		if gd.CustomDeployBlockPath == "" { // Setup paths to use downloaded configs
+			gd.CustomDeployBlockPath = "./deploy_block.txt"
+		} else { // Overriding custom configs, no need to download
+			vlRemoteDpl = false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/generate/generate_scripts.go]: Done [1659f2a](https://github.com/NethermindEth/sedge/pull/196/commits/1659f2ab2246fe933d359c7056e21eb52081c0a0)

File: internal/pkg/generate/generate_scripts.go [hunk 10]
```diff
@@ -180,27 +188,69 @@ func generateDockerComposeScripts(gd GenerationData) (dockerComposePath string,
...
 	if err != nil {
 		return "", err
 	}
+	if ccRemoteGen { // Have to download custom configs
+		if gd.CustomGenesisPath == "" { // Setup paths to use downloaded configs
+			gd.CustomGenesisPath = "./genesis.ssz"
+		} else { // Overriding custom configs, no need to download
+			ccRemoteGen = false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/generate/generate_scripts.go]: Done [1659f2a](https://github.com/NethermindEth/sedge/pull/196/commits/1659f2ab2246fe933d359c7056e21eb52081c0a0)

File: internal/pkg/generate/generate_scripts.go [hunk 11]
```diff
@@ -180,27 +188,69 @@ func generateDockerComposeScripts(gd GenerationData) (dockerComposePath string,
...
 	if err != nil {
 		return "", err
 	}
+	if ccRemoteDpl { // Have to download custom configs
+		if gd.CustomDeployBlockPath == "" { // Setup paths to use downloaded configs
+			gd.CustomDeployBlockPath = "./deploy_block.txt"
+		} else { // Overriding custom configs, no need to download
+			ccRemoteDpl = false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/generate/generate_scripts.go]: Done [1659f2a](https://github.com/NethermindEth/sedge/pull/196/commits/1659f2ab2246fe933d359c7056e21eb52081c0a0)

File: internal/pkg/generate/generate_scripts.go [hunk 12]
```diff
@@ -180,27 +188,69 @@ func generateDockerComposeScripts(gd GenerationData) (dockerComposePath string,
...
 	if err != nil {
 		return "", err
 	}
+	if vlRemoteGen { // Have to download custom configs
+		if gd.CustomGenesisPath == "" { // Setup paths to use downloaded configs
+			gd.CustomGenesisPath = "./genesis.ssz"
+		} else { // Overriding custom configs, no need to download
+			vlRemoteGen = false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/generate/generate_scripts.go]: Done [1659f2a](https://github.com/NethermindEth/sedge/pull/196/commits/1659f2ab2246fe933d359c7056e21eb52081c0a0)

File: internal/pkg/env/get_variable_test.go [hunk 13]
```diff
@@ -78,3 +103,43 @@ func TestGetBootnodes(t *testing.T) {
...
+	t.Parallel()
+
+	tcs := []struct {
+		name         string
+		testdataCase string
+		want         string
+		isErr        bool
+	}{
+		{
+			name:         "Test case 1, no bootnodes",
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/env/get_variable_test.go]: Done [d2369f5](https://github.com/NethermindEth/sedge/pull/196/commits/d2369f581bfdb51e19e5e7023eea36f300a49340)

File: internal/pkg/env/get_variable.go [hunk 14]
```diff
@@ -16,41 +16,101 @@ limitations under the License.
...
+*/
+func GetCCBootnodes(envFilePath string) ([]string, error) {
+	content, err := os.ReadFile(envFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	if m := ReClBOOTNODES.FindStringSubmatch(string(content)); m != nil {
+		m[1] = strings.ReplaceAll(m[1], "\"", "")
+		enodes := strings.Split(m[1], ",")
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply internal/pkg/env/get_variable.go]: Done [9d96ce6](https://github.com/NethermindEth/sedge/pull/196/commits/9d96ce688e9248758098c5717ae98d2556899dd3)

File: templates/services/merge/execution/besu.tmpl (hunk 2 above)
Comment (AntiD2ta) [reply templates/services/merge/execution/besu.tmpl]: Done [8341b27](https://github.com/NethermindEth/sedge/pull/196/commits/8341b27e7c08906845755bfa668a00d602290d25)

File: internal/utils/file_url_utils.go [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/NethermindEth/sedge/configs"
+)
+
+func HandleUrlOrPath(
```
[the remark this answers is not in the document]
Comment (cbermudez97) [reply internal/utils/file_url_utils.go:45]: Done 44916f57978b39af92b8242db31339662f2b8a53

**Review by AntiD2ta — APPROVED** [review]
Approved!

**Review by stdevMac — CHANGES_REQUESTED** [review]

File: cli/cli_utils.go [hunk 16]
```diff
@@ -323,3 +323,73 @@ func preRunTeku(flags *CliCmdFlags) error {
 	}
 	return nil
 }
+
+func LoadCustomNetworksConfig(flags *CliCmdFlags) (string, string, string, string, error) {
```
[the remark this answers is not in the document]
Comment (AntiD2ta) [reply cli/cli_utils.go:327]: `LoadCustomNetworksConfig(flags *CliCmdFlags) (struct, error)`

[the remark this answers is not in the document]
Comment (AntiD2ta) [reply cli/cli_utils.go]: Done [a76ff85](https://github.com/NethermindEth/sedge/pull/196/commits/a76ff8561ee7930365f3d04a9d007ce710e2c139)

**Review by stdevMac — APPROVED** [review]

**Pull request merged**